globalsat_sport.cc

```diff
@@ -0,0 +1,742 @@
+{
+    if (global_opts.debug_level > 1) {
+		printf (MYNAME " serial_deinit()\n");
+	}
+	gbser_deinit (serial_handle);
+	serial_handle = NULL;
+	if (global_opts.debug_level > 1) {
+		printf (MYNAME " serial_deinit() Done\n");
+	}
+}
+
+static int
+serial_recv_byte ()
+{
+	int result;
+	/* allow for a delay of 4s */
+	result = gbser_readc_wait (serial_handle, 4000);
+	switch (result) {
+	case gbser_ERROR:
+		fatal ("serial_recv_byte(): error reading one byte\n");
+		break;
+	case gbser_NOTHING:
+		fatal ("serial_recv_byte(): read timeout\n");
+		break;
+	}
+	return result;
+}
+
+static void
+serial_write_byte (uint8_t byte)
+{
+	int n;
+	if (global_opts.debug_level > 1) {
+		printf ("0x%02x (%d), ", byte, byte);
+	}
+
+	n = gbser_writec (serial_handle, byte);
+	if (n == gbser_ERROR) {
+		fatal ("globalsat_probe_device(): write failed\n");
+	}
+}
+
+
+static void
+globalsat_write_package (uint8_t * payload, uint32_t size)
+{
+	//All globalsat devices but gh561
+	//2 <len_h> <len_l> <payload...> <crc>
+	// gh561 (isSizeSwaped)
+	//2 <len_l> <len_h> <payload...> <crc>
+
+	uint32_t i;
+
+	uint8_t crc = 0;
+	serial_write_byte (2);
+
+	if (!isSizeSwaped) {
+		serial_write_byte ((0xff00 & size) >> 8);
+		crc ^= (0xff00 & size) >> 8;
+		serial_write_byte (0xff & size);
+		crc ^= (0xff & size);
+	}
+	else {
+		serial_write_byte (0xff & size);
+		crc ^= (0xff & size);
+		serial_write_byte ((0xff00 & size) >> 8);
+		crc ^= (0xff00 & size) >> 8;
+	}
+
+	if (payload != NULL) {
+		for (i = 0; i < size; i++) {
+			serial_write_byte (payload[i]);
+			crc ^= payload[i];
+		}
+	}
+	serial_write_byte (crc);
+	if (global_opts.debug_level > 1) {
+		printf ("\n");
+	}
+}
+
+static uint8_t *
+globalsat_read_package (int *out_lenght, uint8_t * out_DeviceCommand)
+{
+	uint8_t DeviceCommand, len_h, len_l, crc;
+	uint8_t *payload;
+	int lenght;
+	int i;
+	uint8_t calc_crc = 0;
+
+	DeviceCommand = serial_recv_byte ();
+	if (global_opts.debug_level > 1) {
+		printf ("DeviceCommand: 0x%02x ", DeviceCommand);
+	}
+	len_h = serial_recv_byte ();
+	calc_crc ^= len_h;
+	len_l = serial_recv_byte ();
+	calc_crc ^= len_l;
+
+	lenght = (len_h << 8) + len_l;
+	if (global_opts.debug_level > 1) {
+		printf ("len=%d Payload:", lenght);
+	}
+
+	payload = (uint8_t *) malloc (lenght);
+	if (payload == NULL) {
+		goto error_out;
+	}
+
+	for (i = 0; i < lenght; i++) {
+		payload[i] = serial_recv_byte ();
+		calc_crc ^= payload[i];
+	}
+
+	crc = serial_recv_byte ();
+	if (global_opts.debug_level > 1) {
+		printf ("crc=0x%x should be=0x%x\n", crc, calc_crc);
+	}
+	if (crc == calc_crc) {
+		*out_DeviceCommand = DeviceCommand;
+		*out_lenght = lenght;
+		return payload;
+	}
+	//crc error
+	free (payload);
+	error_out:
+	*out_lenght = 0;
+	return NULL;
+}
+
+/*
+ * Send on byte package
+ */
+static void
+globalsat_send_simple (uint8_t command)
+{
+	uint8_t payload[1];
+	payload[0] = command;
+	globalsat_write_package (payload, 1);
+}
+
+static void
+globalsat_probe_device ()
+{
+	//2 1 0 191 190
```

Comment:
Some leftover dev comments (it is the bytes set over the USBserial line to the device) Ill just remove them.